File: packages/shared/src/blocking-snapshot.ts
import {
  BLOCKING_PRESETS,
  FULL_FOCUS_ALLOWED_BUNDLE_IDS,
  SOCIAL_MEDIA_BUNDLE_IDS,
  GAME_BUNDLE_IDS,
  SYSTEM_PROTECTED_BUNDLE_IDS,
  appNameForBundleId,
} from './constants';
import type { BlockingPreset, BlockingMode } from './types';

export interface BlockingAppEntry {
  bundleId: string;
  appName: string;
}

/**
 * Immutable record of the blocking policy for a given session, computed once
 * at session start from the class config. Stored as JSONB on `class_sessions`
 * and served verbatim to the iOS / simulator clients.
 */
export interface BlockingSnapshot {
  preset: BlockingPreset;
  mode: BlockingMode;
  blockingActive: boolean;
  blockedApps: BlockingAppEntry[];
  allowedApps: BlockingAppEntry[];
}

function entries(bundleIds: readonly string[]): BlockingAppEntry[] {
  return bundleIds.map((bundleId) => ({
    bundleId,
    appName: appNameForBundleId(bundleId),
  }));
}

/**
 * Resolve a class's blocking config into a session-level snapshot.
 * `customApps` is only consulted when the preset is `'custom'`.
 */
export function resolveBlockingSnapshot(
  preset: BlockingPreset,
  customApps: readonly BlockingAppEntry[] = []
): BlockingSnapshot {
  const protectedSet = new Set<string>(SYSTEM_PROTECTED_BUNDLE_IDS);

  if (preset === BLOCKING_PRESETS.NONE) {
    return {
      preset,
      mode: 'block_specific',
      blockingActive: false,
      blockedApps: [],
      allowedApps: [],
    };
  }

  if (preset === BLOCKING_PRESETS.FULL_FOCUS) {
    return {
      preset,
      mode: 'block_all_except',
      blockingActive: true,
      blockedApps: [],
      allowedApps: entries(FULL_FOCUS_ALLOWED_BUNDLE_IDS),
    };
  }

  if (preset === BLOCKING_PRESETS.NO_SOCIAL_MEDIA) {
    return {
      preset,
      mode: 'block_specific',
      blockingActive: true,
      blockedApps: entries(
        SOCIAL_MEDIA_BUNDLE_IDS.filter((b) => !protectedSet.has(b))
      ),
      allowedApps: [],
    };
  }

  if (preset === BLOCKING_PRESETS.NO_GAMES) {
    return {
      preset,
      mode: 'block_specific',
      blockingActive: true,
      blockedApps: entries(
        GAME_BUNDLE_IDS.filter((b) => !protectedSet.has(b))
      ),
      allowedApps: [],
    };
  }

  // 'custom'
  const dedupedCustom = customApps.filter(
    (a) => !protectedSet.has(a.bundleId)
  );
  return {
    preset,
    mode: 'block_specific',
    blockingActive: dedupedCustom.length > 0,
    blockedApps: dedupedCustom,
    allowedApps: [],
  };
}

/** A snapshot representing "no blocking" — used as a fallback / default. */
export const INACTIVE_BLOCKING_SNAPSHOT: BlockingSnapshot = {
  preset: 'none',
  mode: 'block_specific',
  blockingActive: false,
  blockedApps: [],
  allowedApps: [],
};

File: ios/Bali/Bali/Models/BlockingSnapshot.swift
//
//  BlockingSnapshot.swift
//  Bali — models
//
//  Mirrors `BlockingSnapshot` (packages/shared/src/blocking-snapshot.ts): the
//  immutable, session-frozen blocking policy the server resolves at session
//  start. Bundle IDs / app names here RENDER the policy; iOS enforces via opaque
//  FamilyControls tokens (Phase 7), not these IDs. Unknown preset/mode strings
//  decode to safe fallbacks so a future server value never breaks a screen.
//

import Foundation

nonisolated enum BlockingPreset: String, Codable, Equatable, CaseIterable {
    case none = "none"
    case fullFocus = "full_focus"
    case noSocialMedia = "no_social_media"
    case noGames = "no_games"
    case custom = "custom"

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = BlockingPreset(rawValue: raw) ?? .custom
    }

    var title: String {
        switch self {
        case .none:          return "No Blocking"
        case .fullFocus:     return "Full Focus"
        case .noSocialMedia: return "No Social Media"
        case .noGames:       return "No Games"
        case .custom:        return "Custom"
        }
    }

    /// One-line description used on policy previews / class detail.
    var summary: String {
        switch self {
        case .none:          return "Apps stay unlocked during this class."
        case .fullFocus:     return "Everything but a few essentials is paused."
        case .noSocialMedia: return "Social apps are paused during class."
        case .noGames:       return "Games are paused during class."
        case .custom:        return "A custom set of apps is paused."
        }
    }

    /// SF Symbol for the policy tile / row.
    var icon: String {
        switch self {
        case .none:          return "lock.open.fill"
        case .fullFocus:     return "moon.fill"
        case .noSocialMedia: return "bubble.left.and.bubble.right.fill"
        case .noGames:       return "gamecontroller.fill"
        case .custom:        return "slider.horizontal.3"
        }
    }
}

nonisolated enum BlockingMode: String, Codable, Equatable {
    case blockSpecific = "block_specific"
    case blockAllExcept = "block_all_except"

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = BlockingMode(rawValue: raw) ?? .blockSpecific
    }
}

nonisolated struct BlockingAppEntry: Codable, Equatable, Identifiable, Hashable {
    let bundleId: String
    let appName: String
    var id: String { bundleId }
}

nonisolated struct BlockingSnapshot: Codable, Equatable {
    let preset: BlockingPreset
    let mode: BlockingMode
    let blockingActive: Bool
    let blockedApps: [BlockingAppEntry]
    let allowedApps: [BlockingAppEntry]

    /// "No blocking" fallback (mirrors INACTIVE_BLOCKING_SNAPSHOT).
    static let inactive = BlockingSnapshot(
        preset: .none, mode: .blockSpecific,
        blockingActive: false, blockedApps: [], allowedApps: []
    )

    /// True for an allow-list policy (Full Focus): the meaningful list is the
    /// *allowed* apps; everything else is paused.
    var isAllowList: Bool { mode == .blockAllExcept }

    /// The apps worth showing in the UI for this policy.
    var displayApps: [BlockingAppEntry] {
        isAllowList ? allowedApps : blockedApps
    }
}
